pkg/skaffold/kubernetes/log.go

```diff
@@ -28,132 +29,141 @@ import (
-const defaultRetry int = 5
-
-var colors = []int{
-	31, // red
-	32, // green
-	33, // yellow
-	34, // blue
-	35, // magenta
-	36, // cyan
-	37, // lightGray
-	90, // darkGray
-	91, // lightRed
-	92, // lightGreen
-	93, // lightYellow
-	94, // lightBlue
-	95, // lightPurple
-	96, // lightCyan
-	97, // white
-}
-
 // LogAggregator aggregates the logs for all the deployed pods.
 type LogAggregator struct {
-	muted          int32
-	creationTime   time.Time
-	output         io.Writer
-	retries        int
-	nextColorIndex int
-	lockColor      sync.Mutex
+	muted     int32
+	startTime time.Time
+	output    io.Writer
+
+	images    map[string]int
+	lockImage sync.RWMutex
+
+	containers     map[string]bool
+	lockContainers sync.Mutex
 }
 
 // NewLogAggregator creates a new LogAggregator for a given output.
 func NewLogAggregator(out io.Writer) *LogAggregator {
 	return &LogAggregator{
-		creationTime: time.Now(),
-		output:       out,
-		retries:      defaultRetry,
+		output:     out,
+		images:     map[string]int{},
+		containers: map[string]bool{},
 	}
 }
 
-const streamRetryDelay = 1 * time.Second
+func (a *LogAggregator) Start(ctx context.Context, client corev1.CoreV1Interface) error {
+	a.startTime = time.Now()
 
-// TODO(@r2d4): Figure out how to mock this out. fake.NewSimpleClient
-// won't mock out restclient.Request and will just return a nil stream.
-var getStream = func(r *restclient.Request) (io.ReadCloser, error) {
-	return r.Stream()
-}
+	watcher, err := client.Pods("").Watch(meta_v1.ListOptions{
+		IncludeUninitialized: true,
+	})
+	if err != nil {
+		return err
+	}
 
-func (a *LogAggregator) StreamLogs(client corev1.CoreV1Interface, image string) {
-	for i := 0; i < a.retries; i++ {
-		if err := a.streamLogs(client, image); err != nil {
-			logrus.Infof("Error getting logs %s", err)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case evt := <-watcher.ResultChan():
+				if evt.Type != watch.Added && evt.Type != watch.Modified {
+					continue
+				}
+
+				pod, ok := evt.Object.(*v1.Pod)
+				if !ok {
+					continue
+				}
+
+				color, present := a.podColor(pod)
+				if present {
+					go a.streamLogs(client, pod, color)
+				}
+			}
 		}
-		time.Sleep(streamRetryDelay)
-	}
+	}()
+
+	return nil
 }
 
-func (a *LogAggregator) SetCreationTime(t time.Time) {
-	a.creationTime = t
+// RegisterImage register an image for which pods need to be logged.
+func (a *LogAggregator) RegisterImage(image string, color int) {
+	a.lockImage.Lock()
+	a.images[image] = color
+	a.lockImage.Unlock()
 }
 
-// nolint: interfacer
-func (a *LogAggregator) streamLogs(client corev1.CoreV1Interface, image string) error {
-	pods, err := client.Pods("").List(meta_v1.ListOptions{
-		IncludeUninitialized: true,
-	})
-	if err != nil {
-		return errors.Wrap(err, "getting pods")
+// podColor reads the images used by a pod and compares them to the list
+// of registered colors.
+func (a *LogAggregator) podColor(pod *v1.Pod) (color int, found bool) {
```

Comment:
ideally log color wouldn't be so tightly coupled with actually choosing which logs to stream, but I think this is OK for now.